app/components/elements/form/input/index.js

```diff
@@ -0,0 +1,71 @@
+import {reject} from 'wasmuth'
+import {dispatch} from '/store'
+import BaseTextInput from './base'
+import {compose, setNodeName} from '/util/compose'
+import {
+  formSet as formSetBuilder,
+  formRemove as formRemoveBuilder
+} from '../actions.js'
+import {fieldState} from '../selectors.js'
+
+export default compose(setNodeName('TextInput'), {
+  componentWillMount () {
+    this.formSet = formSetBuilder(this.props.formName)
+    this.formRemove = formRemoveBuilder(this.props.formName)
+  },
+  render ({name, className = '', formSet, formRemove, formName, type = 'text', ...props}) {
+    const {value, error, asyncError, touched, focus, modified} = fieldState(
+      formName,
+      name
+    )
+    return BaseTextInput({
+      value,
+      className: `${!!(error || asyncError) && !!touched && !focus && 'error'} ${className}`,
+      onFocus: ev => {
+        !touched && dispatch(formSet(['touched', name], true))
```

Comment:
I wonder if these should be opt in? I've already had to manage perf issues on Qee with just syncing input changes. I could see large forms with all these events/state changes being a potential perf issue.